core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -32,6 +32,9 @@ private[spark] sealed trait MapStatus {
   /** Location where this task was run. */
   def location: BlockManagerId
 
+  /** stage attempt for the ShuffleMapTask */
+  def stageAttemptId: Int
```

Comment:
`MapStatus` is now needs to keep a `stageAttemptId` so the fetch side knows which stage attempt to read data from.  This is a negligible increase to the size of the map status when we have just one stage attempt, since its just an extra int.  But when there are multiple attempts, we now have one `MapStatus` per (location, stageAttempt).  In pathalogical cases, this will result in many more `MapStatus`.

I needed to update the serialization of `MapStatus`, but I believe this should be covered by existing tests.